src/main/scala/util/BundleMap.scala

```diff
@@ -37,6 +44,81 @@ class BundleMap(val fields: Seq[BundleField]) extends Record {
   require(fields.map(_.key.name).distinct.size == fields.size)
   override def cloneType: this.type = (new BundleMap(fields)).asInstanceOf[this.type]
   val elements = ListMap(fields.map { bf => bf.key.name -> chisel3.experimental.DataMirror.internal.chiselTypeClone(bf.data) } :_*)
-  def apply[T <: Data](key: BundleKey[T]) = elements(key.name).asInstanceOf[T]
+
+  def apply[T <: Data](key: BundleKey[T]): T         = elements(key.name).asInstanceOf[T]
+  def lift [T <: Data](key: BundleKey[T]): Option[T] = elements.lift(key.name).map(_.asInstanceOf[T])
+
+  def apply(key: BundleKeyBase): Data         = elements(key.name)
+  def lift (key: BundleKeyBase): Option[Data] = elements.lift(key.name)
+
+  // For every key where this.output or that.input
+  def assignL(that: BundleMap): Seq[BundleKeyBase] = {
+    val a = this.fields.filter( _.isOutput).map(_.key)
+    val b = that.fields.filter(!_.isOutput).map(_.key)
+    (a ++ b).toList.distinct
+  }
+
+  // For every key where this.input or that.output
+  def assignR(that: BundleMap): Seq[BundleKeyBase] = {
+    val a = this.fields.filter(!_.isOutput).map(_.key)
+    val b = that.fields.filter( _.isOutput).map(_.key)
+    (a ++ b).toList.distinct
+  }
 }
 
+// Implement :<= :=> and :<>
+object FixChisel3 {
+  private def descendL(x: Data, y: Data): Unit = {
+    DataMirror.specifiedDirectionOf(x) match {
+      case SpecifiedDirection.Unspecified => assignL(x, y)
+      case SpecifiedDirection.Output      => x := y
+      case SpecifiedDirection.Input       => ()
+      case SpecifiedDirection.Flip        => assignR(y, x)
+    }
+  }
+
+  private def descendR(x: Data, y: Data): Unit = {
+    DataMirror.specifiedDirectionOf(y) match {
+      case SpecifiedDirection.Unspecified => assignR(x, y)
+      case SpecifiedDirection.Output      => ()
+      case SpecifiedDirection.Input       => y := x
+      case SpecifiedDirection.Flip        => assignL(y, x)
+    }
+  }
+
+  def assignL(x: Data, y: Data): Unit = {
```

Comment:
If this PR is going to `FixChisel3`, did you consider returning `x` here instead of `Unit`?

That has seemed to be one of the nice behaviors of the diplomatic connectors that might be worth exploring here.

Reply:
That's an interesting idea. I'm not sure it's that useful in this context, though. In diplomacy, when you go `a := b := c`, generally `b` does something between the two connections. In chisel, this sort of expression would only be legal if 'b' were a wire, and thus it's a no-op.

Reply:
I think I've wanted this for constructs like the following where you're iterating through something connecting things as you go along. I realize this effectively saves one line, but it's just always seemed obnoxious that we return `Unit` when I'm not sure there's a good reason for it. Granted `def foo { val x = 3 }` is of type `Unit`.

```scala
class ShiftRegister(a: Int) extends MultiIOModule {

  val in: Bool = IO(Input(Bool()))
  val out = IO(Output(Bool()))

  val state: Seq[Bool] = Seq.fill(a)(Reg(Bool()))

  out := state.foldLeft(in){ case (right, left) => left := right; left }

}
```

Reply:
Ah. So your point is it's not just legal for wires, but also registers. I still think that until/unless chisel3 supports connecting circuits with ':=' instead of just Data, this is going to have very limited utility. Basically, it's only good for shift registers.